packages/realm_dart/lib/src/native/from_native.dart

```diff
@@ -0,0 +1,358 @@
+    if (this == nullptr || length == 0) {
+      return null;
+    }
+    List<CompensatingWriteInfo> compensatingWrites = [];
+    for (int i = 0; i < length; i++) {
+      final compensatingWrite = this[i];
+      final reason = compensatingWrite.reason.cast<Utf8>().toDartString();
+      final objectName = compensatingWrite.object_name.cast<Utf8>().toDartString();
+      final primaryKey = compensatingWrite.primary_key.toPrimitiveValue();
+      compensatingWrites.add(CompensatingWriteInfo(objectName, reason, RealmValue.from(primaryKey)));
+    }
+    return compensatingWrites;
+  }
+}
+
+extension PointerRealmErrorEx on Pointer<realm_error_t> {
+  SyncError toDart() {
+    final message = ref.message.cast<Utf8>().toDartString();
+    final details = SyncErrorDetails(message, ref.error, ref.user_code_error.toUserCodeError());
+    return SyncErrorInternal.createSyncError(details);
+  }
+}
+
+extension ObjectEx on Object {
+  Pointer<Void> toPersistentHandle() {
+    return realmLib.realm_dart_object_to_persistent_handle(this);
+  }
+}
+
+extension ListUserStateEx on List<UserState> {
+  UserState fromIndex(int index) {
+    if (!UserState.values.any((value) => value.index == index)) {
+      throw RealmError("Unknown user state $index");
+    }
+
+    return UserState.values[index];
+  }
+}
+
+extension RealmPropertyInfoEx on realm_property_info {
+  SchemaProperty toSchemaProperty() {
+    final linkTarget = link_target == nullptr ? null : link_target.cast<Utf8>().toDartString();
+    return SchemaProperty(name.cast<Utf8>().toDartString(), RealmPropertyType.values[type],
+        optional: flags & realm_property_flags.RLM_PROPERTY_NULLABLE == realm_property_flags.RLM_PROPERTY_NULLABLE,
+        primaryKey: flags & realm_property_flags.RLM_PROPERTY_PRIMARY_KEY == realm_property_flags.RLM_PROPERTY_PRIMARY_KEY,
+        linkTarget: linkTarget == null || linkTarget.isEmpty ? null : linkTarget,
+        collectionType: RealmCollectionType.values[collection_type]);
+  }
+}
+
+extension CompleterEx<T> on Completer<T> {
+  void completeFrom(FutureOr<T> Function() action) {
+    try {
+      complete(action());
+    } catch (error, stackTrace) {
+      completeError(error, stackTrace);
+    }
+  }
+
+  void completeWithAppError(Pointer<realm_app_error> error) {
+    final message = error.ref.message.cast<Utf8>().toRealmDartString()!;
+    final linkToLogs = error.ref.link_to_server_logs.cast<Utf8>().toRealmDartString();
+    completeError(AppInternal.createException(message, linkToLogs, error.ref.http_status_code));
+  }
+}
+
+enum CustomErrorCode {
+  noError(0),
+  unknownHttp(998),
+  unknown(999),
+  timeout(1000);
+
+  final int code;
+  const CustomErrorCode(this.code);
+}
+
+enum HttpMethod {
+  get,
+  post,
+  patch,
+  put,
+  delete,
+}
+
+extension RealmTimestampEx on realm_timestamp_t {
+  DateTime toDart() {
+    return DateTime.fromMicrosecondsSinceEpoch(seconds * 1000000 + nanoseconds ~/ 1000, isUtc: true);
```

Comment:
May improve some readability there 👍 (Unless Dart has a number delimiter which I didn't find from some quick DuckDuckGoing.)
```suggestion
    return DateTime.fromMicrosecondsSinceEpoch(seconds * 1000 * 1000 + nanoseconds ~/ 1000, isUtc: true);
```

Reply:
I'll use the `_microsecondsPerSecond` constant which is even more readable